### PR TITLE
Declarations+jsDoc for Enumerator, VBArray

### PR DESCRIPTION
**Pull request by zspitz**
Addresses #2740

### CODE REVIEW

File: src/lib/scriptHost.d.ts [hunk 1]
```diff
@@ -170,3 +170,64 @@ declare var WScript: {
      */
     Sleep(intTime: number): void;
 };
+
+/*
+ * Allows enumerating over a COM collection, which may not have indexed item access
+ */
+interface Enumerator<T> {
+    /*
```
Comment (mhegazy) [inline src/lib/scriptHost.d.ts]: doc should start with /**

File: src/lib/scriptHost.d.ts [hunk 2]
```diff
@@ -170,3 +170,64 @@ declare var WScript: {
...
+     * Resets the current item in the collection to the first item. If there are no items in the collection, the current item is set to undefined.
+     */
+    moveFirst(): void;
+    /*
+     * Moves the current item to the next item in the collection. If the enumerator is at the end of the collection or the collection is empty, the current item is set to undefined.
+     */
+    moveNext(): void;
+}
+interface EnumeratorConstructor {
```
Comment (mhegazy) [inline src/lib/scriptHost.d.ts]: new lines please

Comment (DanielRosenwasser) [inline src/lib/scriptHost.d.ts]: Between all declarations, to be clear

File: src/lib/scriptHost.d.ts [hunk 3]
```diff
@@ -170,3 +170,64 @@ declare var WScript: {
...
+    /*
+     * Returns the current item in the collection
+     */
+    item(): T;
+    /*
+     * Resets the current item in the collection to the first item. If there are no items in the collection, the current item is set to undefined.
+     */
+    moveFirst(): void;
+    /*
+     * Moves the current item to the next item in the collection. If the enumerator is at the end of the collection or the collection is empty, the current item is set to undefined.
```
Comment (DanielRosenwasser) [inline src/lib/scriptHost.d.ts]: Wrap

File: src/lib/scriptHost.d.ts [hunk 4]
```diff
@@ -170,3 +170,64 @@ declare var WScript: {
...
+     * @param dimension 1-based dimension (defaults to 1)
+     */
+    lbound(dimension?: number): number;
+    /*
+     * Returns the largest available index for a given dimension
+     * @param dimension 1-based dimension (defaults to 1)
+     */
+    ubound(dimension?: number): number;
+    /*
+     * Returns a Javascript array with all the elements in the VBArray. If there are multiple dimensions, each successive dimension is appended to the end of the array.
```
Comment (DanielRosenwasser) [inline src/lib/scriptHost.d.ts]: Wrap

File: src/lib/scriptHost.d.ts [hunk 5]
```diff
@@ -170,3 +170,64 @@ declare var WScript: {
...
+    /*
+     * Returns true if the current item is the last one in the collection, or the collection is empty, or the current item is undefined
+     */
+    atEnd(): boolean;
+    /*
+     * Returns the current item in the collection
+     */
+    item(): T;
+    /*
+     * Resets the current item in the collection to the first item. If there are no items in the collection, the current item is set to undefined.
```
Comment (DanielRosenwasser) [inline src/lib/scriptHost.d.ts]: Wrap

File: src/lib/scriptHost.d.ts [hunk 6]
```diff
@@ -170,3 +170,64 @@ declare var WScript: {
      */
     Sleep(intTime: number): void;
 };
+
+/*
+ * Allows enumerating over a COM collection, which may not have indexed item access
+ */
+interface Enumerator<T> {
+    /*
+     * Returns true if the current item is the last one in the collection, or the collection is empty, or the current item is undefined
```
Comment (DanielRosenwasser) [inline src/lib/scriptHost.d.ts]: Wrap

**Comment by zspitz** [comment]
@DanielRosenwasser When wrapping, how many characters per line, more or less?

**Comment by mhegazy** [comment]
i usually do 100

**Comment by zspitz** [comment]
@DanielRosenwasser  @mhegazy RE: newlines -- before each member, or only before each type/declaration (`interface`,`declare var`)?

File: src/lib/scriptHost.d.ts [hunk 7]
```diff
@@ -88,85 +99,180 @@ interface TextStreamReader extends TextStreamBase {
 
 declare var WScript: {
     /**
-    * Outputs text to either a message box (under WScript.exe) or the command console window followed by a newline (under CScript.ext).
+    * Outputs text to either a message box (under WScript.exe) or the command console window followed by
+    * a newline (under CScript.ext).
```
Comment (DanielRosenwasser) [inline src/lib/scriptHost.d.ts]: ext->exe

File: src/lib/scriptHost.d.ts [hunk 8]
```diff
@@ -88,85 +99,180 @@ interface TextStreamReader extends TextStreamBase {
...
     /**
      * Disconnects a COM object from its event sources.
      */
     DisconnectObject(obj: any): void;
+
     /**
      * Retrieves an existing object with the specified ProgID from memory, or creates a new one from a file.
-     * @param strPathname Fully qualified path to the file containing the object persisted to disk. For objects in memory, pass a zero-length string.
+     * @param strPathname Fully qualified path to the file containing the object persisted to disk.
+     * For objects in memory, pass a zero-length string.
```
Comment (DanielRosenwasser) [inline src/lib/scriptHost.d.ts]: Indent the word "For" a few times until it passes the word "Fully"

**Comment by DanielRosenwasser** [comment]
The spacing looks perfect @zspitz.

**Comment by zspitz** [comment]
Fixed @DanielRosenwasser  @mhegazy

**Comment by DanielRosenwasser** [comment]
:+1:

**Comment by mhegazy** [comment]
Thanks!

**Pull request merged**